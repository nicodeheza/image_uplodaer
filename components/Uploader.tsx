export default function Uploader() {
	return (
		<div data-testid="Uploader_container" style={{maxWidth: "402px", maxHeight: "469px"}}>
			<div></div>
		</div>
	);
}
