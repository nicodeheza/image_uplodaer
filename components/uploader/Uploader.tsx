import Image from "next/image";
import {useRef, useState, Dispatch, SetStateAction} from "react";
import styles from "./Uploader.module.css";

export interface UploaderProps {
	setFile: Dispatch<SetStateAction<undefined>>;
}

export default function Uploader({setFile}: UploaderProps) {
	const [dragActive, setDragActive] = useState(false);

	const inputRef = useRef<null | HTMLInputElement>(null);

	function isImage(fileType: string) {
		const regEx = /^image\/+[\w\W]+/;
		const is = regEx.test(fileType);
		if (!is) alert("Only image files are accepted");
		return is;
	}

	const handleDrag = function (e: any) {
		e.preventDefault();
		e.stopPropagation();
		if (e.type === "dragenter" || e.type === "dragover") {
			setDragActive(true);
		} else if (e.type === "dragleave") {
			setDragActive(false);
		}
	};

	const handleDrop = function (e: any) {
		e.preventDefault();
		e.stopPropagation();
		setDragActive(false);
		if (e.dataTransfer.files && e.dataTransfer.files[0]) {
			const fileType = e.dataTransfer.files[0].type;
			if (isImage(fileType)) {
				setFile(e.dataTransfer.files[0]);
			}
		}
	};

	const handleChange = function (e: any) {
		e.preventDefault();
		if (e.target.files && e.target.files[0]) {
			const fileType = e.target.files[0].type;
			if (isImage(fileType)) {
				setFile(e.target.files[0]);
			}
		}
	};
	const onButtonClick = () => {
		inputRef.current?.click();
	};

	return (
		<form
			id="form-file-upload"
			onDragEnter={handleDrag}
			onSubmit={(e) => e.preventDefault()}
			title="Image form"
			className={styles.formCard}
		>
			<h1>Upload your image</h1>
			<p className={styles.primaryText}>File should be Jpeg, Png,...</p>
			<input
				ref={inputRef}
				type="file"
				id="input-file-upload"
				multiple={true}
				onChange={handleChange}
				title="file input"
				accept="image/*"
			/>
			<label id="label-file-upload" htmlFor="input-file-upload">
				<div
					className={styles.dropView}
					style={dragActive ? {border: "solid 1px red"} : {}}
				>
					<Image width={114.13} height={88.24} src="/image.svg" alt="image" priority />
					<p className={styles.secondaryText}>Drag & Drop your image here</p>
				</div>
			</label>
			<p className={styles.secondaryText + " " + styles.or}>Or</p>
			<button onClick={onButtonClick}>Choose a file</button>
			{dragActive && (
				<div
					id="drag-file-element"
					data-testid="D&D"
					onDragEnter={handleDrag}
					onDragLeave={handleDrag}
					onDragOver={handleDrag}
					onDrop={handleDrop}
					className={styles.drop}
				></div>
			)}
		</form>
	);
}
