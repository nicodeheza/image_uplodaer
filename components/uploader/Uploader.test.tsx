import Uploader, {UploaderProps} from "./Uploader";
import {fireEvent, render} from "@testing-library/react";
import renderer from "react-test-renderer";

describe("Uploader component", () => {
	it("update setFile when image is drag and drop to the D&D container", () => {
		const setFile = jest.fn();
		const files = [new File(["(⌐□_□)"], "chucknorris.png", {type: "image/png"})];
		const {getByTestId, getByTitle} = render(<Uploader setFile={setFile} />);
		fireEvent.dragEnter(getByTitle("Image form"));
		fireEvent.drop(getByTestId("D&D"), {
			dataTransfer: {
				files
			}
		});
		expect(setFile.mock.calls.length).toBe(1);
		expect(setFile.mock.calls[0]).toEqual(files);
	});
	it("when button is clicked  select an image from a folder", () => {
		const setFile = jest.fn();
		const files = [new File(["(⌐□_□)"], "chucknorris.png", {type: "image/png"})];
		const {getByText, getByTitle} = render(<Uploader setFile={setFile} />);
		const input = getByTitle("file input");
		const inputOnClickSpy = jest.spyOn(input, "click");

		fireEvent.click(getByText("Choose a file"));

		expect(inputOnClickSpy).toBeCalled();

		fireEvent.change(input, {
			target: {
				files
			}
		});

		expect(setFile.mock.calls.length).toBe(1);
		expect(setFile.mock.calls[0]).toEqual(files);
	});
	it("if trying to upload non-image file, alert user", () => {
		const setFile = jest.fn();
		const files = [new File(["(⌐□_□)"], "chucknorris.html", {type: "text/html"})];
		const {getByText, getByTitle} = render(<Uploader setFile={setFile} />);
		const input = getByTitle("file input");
		const saveAlert = window.alert;
		window.alert = jest.fn();

		fireEvent.change(input, {
			target: {
				files
			}
		});

		expect(setFile.mock.calls.length).toBe(0);
		expect(window.alert).toHaveBeenCalled();
		window.alert = saveAlert;
	});
	it("snapshot", () => {
		const setFile = jest.fn();
		const tree = renderer.create(<Uploader setFile={setFile} />).toJSON();
		expect(tree).toMatchSnapshot();
	});
});
