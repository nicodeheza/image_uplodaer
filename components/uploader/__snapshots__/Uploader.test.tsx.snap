// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Uploader component snapshot 1`] = `
<form
  className="formCard"
  id="form-file-upload"
  onDragEnter={[Function]}
  onSubmit={[Function]}
  title="Image form"
>
  <h1>
    Upload your image
  </h1>
  <p
    className="primaryText"
  >
    File should be Jpeg, Png,...
  </p>
  <input
    accept="image/*"
    id="input-file-upload"
    multiple={true}
    onChange={[Function]}
    title="file input"
    type="file"
  />
  <label
    htmlFor="input-file-upload"
    id="label-file-upload"
  >
    <div
      className="dropView"
      style={Object {}}
    >
      <span
        style={
          Object {
            "background": "none",
            "border": 0,
            "boxSizing": "border-box",
            "display": "inline-block",
            "height": "initial",
            "margin": 0,
            "maxWidth": "100%",
            "opacity": 1,
            "overflow": "hidden",
            "padding": 0,
            "position": "relative",
            "width": "initial",
          }
        }
      >
        <span
          style={
            Object {
              "background": "none",
              "border": 0,
              "boxSizing": "border-box",
              "display": "block",
              "height": "initial",
              "margin": 0,
              "maxWidth": "100%",
              "opacity": 1,
              "padding": 0,
              "width": "initial",
            }
          }
        >
          <img
            alt=""
            aria-hidden={true}
            src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%27114.13%27%20height=%2788.24%27/%3e"
            style={
              Object {
                "background": "none",
                "border": 0,
                "display": "block",
                "height": "initial",
                "margin": 0,
                "maxWidth": "100%",
                "opacity": 1,
                "padding": 0,
                "width": "initial",
              }
            }
          />
        </span>
        <img
          alt="image"
          data-nimg="intrinsic"
          decoding="async"
          onError={[Function]}
          onLoad={[Function]}
          src="/image.svg"
          srcSet="/image.svg 1x, /image.svg 2x"
          style={
            Object {
              "border": "none",
              "bottom": 0,
              "boxSizing": "border-box",
              "display": "block",
              "height": 0,
              "left": 0,
              "margin": "auto",
              "maxHeight": "100%",
              "maxWidth": "100%",
              "minHeight": "100%",
              "minWidth": "100%",
              "objectFit": undefined,
              "objectPosition": undefined,
              "padding": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "width": 0,
            }
          }
        />
      </span>
      <p
        className="secondaryText"
      >
        Drag & Drop your image here
      </p>
    </div>
  </label>
  <p
    className="secondaryText or"
  >
    Or
  </p>
  <button
    onClick={[Function]}
  >
    Choose a file
  </button>
</form>
`;
